hibernate-rx-core/src/test/java/org/hibernate/rx/CascadeComplicatedTest.java

```diff
@@ -0,0 +1,442 @@
+package org.hibernate.rx;
+
+import io.vertx.ext.unit.TestContext;
+import org.hibernate.cfg.Configuration;
+import org.hibernate.Hibernate;
+import org.junit.Before;
+import org.junit.Test;
+
+import javax.persistence.*;
+import java.util.Set;
+import java.util.UUID;
+import java.util.concurrent.CompletionStage;
+
+/**
+ * This test uses a complicated model that requires Hibernate to delay
+ * inserts until non-nullable transient entity dependencies are resolved.
+ *
+ * All IDs are generated from a sequence.
+ *
+ * JPA cascade types are used (javax.persistence.CascadeType)..
+ *
+ * This test uses the following model:
+ *
+ * <code>
+ *     ------------------------------ N G
+ *     |
+ *     |                                1
+ *     |                                |
+ *     |                                |
+ *     |                                N
+ *     |
+ *     |         E N--------------0,1 * F
+ *     |
+ *     |         1                      N
+ *     |         |                      |
+ *     |         |                      |
+ *     1         N                      |
+ *     *                                |
+ *     B * N---1 D * 1------------------
+ *     *
+ *     N         N
+ *     |         |
+ *     |         |
+ *     1         |
+ *               |
+ *     C * 1-----
+ *</code>
+ *
+ * In the diagram, all associations are bidirectional;
+ * assocations marked with '*' cascade persist, save, merge operations to the
+ * associated entities (e.g., B cascades persist to D, but D does not cascade
+ * persist to B);
+ *
+ * b, c, d, e, f, and g are all transient unsaved that are associated with each other.
+ *
+ * When persisting b with ORM, the entities are added to the ActionQueue in the following order:
+ * c, d (depends on e), f (depends on d, g), e, b, g.
```

Comment:
`session.persist(b)` adds insert actions to `RxActionQueue` in the same order as documented above.

When `b` and all its associated entities are transient, `session.merge(b)` adds insert actions to `RxActionQueue` in a different order, and 2 `D` entities get added(?!?): `D (id ==2)`,  `c`, `f`, `D (id ==4)`, `e`, `b`, `g`.